File: ColorPicker/DTColorPickerImageView.m
//
//  DTColorPickerImageView.m
//  ColorPicker
//
//  Created by Darktt on 2015/3/19.
//  Copyright (c) 2015年 Darktt. All rights reserved.
//

#import "DTColorPickerImageView.h"
#import "UIImage+ColorPicker.h"

@interface DTColorPickerImageView()
{
    DTColorPickerHandler _handler;
}

@end

@implementation DTColorPickerImageView

+ (instancetype)colorPickerWithFrame:(CGRect)frame
{
    DTColorPickerImageView *colorPicker = [[DTColorPickerImageView alloc] initWithFrame:frame];
    
    return [colorPicker autorelease];
}

+ (instancetype)colorPickerWithImage:(UIImage *)image
{
    DTColorPickerImageView *colorPicker = [[DTColorPickerImageView alloc] initWithImage:image];
    
    return [colorPicker autorelease];
}

#pragma mark - Instance Methods

- (void)awakeFromNib
{
    [super awakeFromNib];
    
    [self setUserInteractionEnabled:YES];
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self == nil) return nil;
    
    [self setUserInteractionEnabled:YES];
    
    return self;
}

- (instancetype)initWithImage:(UIImage *)image
{
    self = [super initWithImage:image];
    if (self == nil) return nil;
    
    [self setUserInteractionEnabled:YES];
    
    return self;
}

- (void)dealloc
{
    if (_handler != nil) {
        Block_release(_handler);
    }
    
    [super dealloc];
}

- (void)handlesDidPickColor:(DTColorPickerHandler)handler
{
    if (_handler != nil) {
        Block_release(_handler);
    }
    
    _handler = Block_copy(handler);
}

#pragma mark - Touch events

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint location = [touch locationInView:self];
    
    [self pickerColorAtPoint:location];
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    CGPoint location = [touch locationInView:self];
    
    // Respondes when point in self bounds.
    if (!CGRectContainsPoint(self.bounds, location)) {
        return;
    }
    
    [self pickerColorAtPoint:location];
}

- (void)pickerColorAtPoint:(CGPoint)point
{
    CGPoint convertPoint = [self.image convertPoint:point fromImageView:self];
    
    UIColor *color = [self.image pickColorWithPoint:convertPoint];
    
    if (_handler != nil) {
        _handler(color);
        
        return;
    }
    
    if ([self.delegate respondsToSelector:@selector(imageView:didPickColorWithColor:)]) {
        [self.delegate imageView:self didPickColorWithColor:color];
    }
}

@end

File: DTColorPicker/DTColorPickerImageView.swift
//
//  DTColorPickerImageView.swift
//
//  Created by Darktt on 15/11/20.
//  Copyright © 2015年 Darktt. All rights reserved.
//

import UIKit

public typealias DTColorPickerHandler = (_ color: UIColor) -> Void

@objc
public protocol DTColorPickerImageViewDelegate: NSObjectProtocol
{
    @objc 
    optional func imageView(_ imageView: DTColorPickerImageView, didPickColorWithColor color: UIColor) -> Void
}

public class DTColorPickerImageView: UIImageView
{
    @IBOutlet weak public var delegate: DTColorPickerImageViewDelegate?
    
    fileprivate var handler: DTColorPickerHandler?
    
    public override func awakeFromNib()
    {
        super.awakeFromNib()
        
        self.isUserInteractionEnabled = true
    }
    
    public override init(frame: CGRect)
    {
        super.init(frame: frame)
        
        self.isUserInteractionEnabled = true
    }
    
    public override init(image: UIImage?)
    {
        super.init(image: image)
        
        self.isUserInteractionEnabled = true
    }
    
    private override init(image: UIImage?, highlightedImage: UIImage?)
    {
        super.init(image: image, highlightedImage: highlightedImage)
        
        self.isUserInteractionEnabled = true
    }

    public required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
    }
    
    /**
     When handler and delegate not nil, will respondes handler first.
     
     - parameter handler: The hander for respondes.
     */
    public func handlesDidPickColor(handler: DTColorPickerHandler?) -> Void
    {
        self.handler = handler
    }
    
    //MARK: - Touch events
    
    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?)
    {
        guard let touch: UITouch = touches.first else {
            return
        }
        
        let location: CGPoint = touch.location(in: self)
        
        self.pickerColor(at: location)
    }
    
    public override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?)
    {
        guard let touch: UITouch = touches.first else {
            return
        }
        
        let location: CGPoint = touch.location(in: self)
        
        // Respondes when point in self bounds.
        if (!self.bounds.contains(location)) {
            return
        }
        
        self.pickerColor(at: location)
    }
}

// MARK: - Private Methods -

fileprivate extension DTColorPickerImageView
{
    // MARK: Private Method
    
    func pickerColor(at point: CGPoint) -> Void
    {
        guard let image: UIImage = self.image else {
            return
        }
        
        let convertPoint: CGPoint = image.convertPoint(point, from: self)
        
        let color: UIColor = image.pickColor(at: convertPoint)
        
        if let handler = self.handler {
            handler(color)
            return
        }
        
        if let imageViewMethod = self.delegate?.imageView {
            
            imageViewMethod(self, color)
        }
    }
}
